pkg/lib/cockpit.js

```diff
@@ -2330,24 +2331,24 @@ function factory() {
                 if (watch_channel)
                     return;
 
-                const opts = {
-                    payload: "fswatch1",
-                    path,
-                    superuser: base_channel_options.superuser,
-                };
-                watch_channel = cockpit.channel(opts);
-                watch_channel.addEventListener("message", function (event, message_string) {
-                    let message;
-                    try {
-                        message = JSON.parse(message_string);
-                    } catch (e) {
-                        message = null;
-                    }
-                    if (message && message.path == path && message.tag && message.tag != watch_tag) {
-                        if (options && options.read !== undefined && !options.read)
-                            fire_watch_callbacks(null, message.tag);
-                        else
-                            read();
+                watch_channel = new FsInfoClient(path, ["tag"], { superuser: base_channel_options.superuser });
+                watch_channel.on('change', (state) => {
+                    if (state.error) {
+                        // Behave like fsread1, not-found is not a fatal error
+                        if (state.error.problem === "not-found") {
+                            fire_watch_callbacks(null, "-");
+                        } else {
+                            const error = new BasicError(state.error.problem, state.error.message);
+                            fire_watch_callbacks(null, "-", error);
+                        }
+                    } else if (state.info && state.info.tag) {
+                        // otherwise, the file is present with the given tag
+                        if (state.info.tag !== watch_tag) {
+                            if (options?.read)
```

Comment:
Mea culpa: this will default to read: false, which is a behaviour change.  I think that your previous check if `options?.read === false` was the correct one (but with the if/else the other way around).